Add support for endianness

Hogflash opened the pull request:
Filip, this is my first PR ever, so let me know if I am doing things correctly or you have any questions/comments.  There are a few new files, and changes to the main code to access the endian and address display format properties.

fjeremic commented on HexView.Wpf/ExtensionMethods.cs:
```diff
@@ -0,0 +1,129 @@
+        /// </summary>
+        /// <param name="reader">in-use BinaryReader.</param>
+        /// <returns>UInt64 (unsigned long).</returns>
+        public static ulong ReadUInt64BE(this BinaryReader reader)
+        {
+            var data = reader.ReadBytes(8);
+            Array.Reverse(data);
+            return BitConverter.ToUInt64(data, 0);
+        }
```
This implementation is going to be quite slow with the use of arrays. When rendering large amounts of data this will have noticeable effects. In one of my other projects I actually have an `EndianBinaryReader` already implemented. You can find it here:

https://github.com/fjeremic/BinaryTools.Elf/blob/master/BinaryTools.Elf/Io/EndianBinaryReader.cs

Actually in that directory you will find much of the same APIs as in your PR:
https://github.com/fjeremic/BinaryTools.Elf/tree/master/BinaryTools.Elf/Io

I'm wondering if we could just pull from there. For example your `ReadUInt64BE` is implemented here:
https://github.com/fjeremic/BinaryTools.Elf/blob/master/BinaryTools.Elf/Io/EndianBinaryReader.cs#L153

Which delegates to this call:
https://github.com/fjeremic/BinaryTools.Elf/blob/master/BinaryTools.Elf/Io/EndianBitConverter.cs#L94

The reason this is much faster is because the JIT compiler will recognize this pattern and turn it into `bswap` x86 instruction for example which is as fast as it gets.

Do you think it is feasible to borrow from the aforementioned project?

Hogflash replied:
Filip, please use your faster code for this.  See additional comment below. (I didn't see spot for commenting here...)

fjeremic commented:
> Filip, this is my first PR ever, so let me know if I am doing things correctly or you have any questions/comments. There are a few new files, and changes to the main code to access the endian and address display format properties.

Thanks! Yes this mostly looks good, and your PR is written in a way that follows existing conventions. Cheers!

Hogflash commented:
Happy to use your endian bit converter code, performance is always a good thing!  Please mod code as needed!

--Gary

fjeremic commented:
I looked over it again and I think the change is valuable enough to merge. I can make the performance improvements in a subsequent commit. I'll leave the above review unresolved and open an issue against it so we don't forget about it. Thanks so much for the contribution @Hogflash. Cheers!

Hogflash commented:
Filip, I spent some time looking at the endian issue, and after some investigation and testing, I updated the extension methods to use your shift-or extraction methods, which are about 2.5x faster than using the array.reverse or binaryPrimitives  code.  I added a test program used for checking performance of the endian extraction.

I also added little endian routines which are needed if the computer architecture is big endian.  Hopefully this covers all possibilities, and the code looks OK to you.
